substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/ThreadMXUtils.java

```diff
@@ -0,0 +1,176 @@
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.svm.core.jdk;
+
+import com.oracle.svm.core.SubstrateUtil;
+import com.oracle.svm.core.jdk.management.SubstrateThreadMXBean;
+import com.oracle.svm.core.locks.Target_java_util_concurrent_locks_AbstractOwnableSynchronizer;
+import com.oracle.svm.core.monitor.JavaMonitor;
+import com.oracle.svm.core.thread.JavaThreads.JMXMonitoring;
+import com.oracle.svm.core.thread.PlatformThreads;
+import org.graalvm.collections.EconomicSet;
+
+import java.lang.management.ThreadInfo;
+import java.util.ArrayDeque;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.concurrent.locks.AbstractOwnableSynchronizer;
+import java.util.concurrent.locks.LockSupport;
+
+/**
+ * Utils to support {@link SubstrateThreadMXBean} for providing threading information for JMX
+ * support. Include the {@link ThreadInfo} constructing utils, and deadlock detection utils.
+ */
+public class ThreadMXUtils {
+
+    public static class ThreadInfoConstructionUtils {
+
+        private static StackTraceElement[] getStackTrace(Thread thread, int maxDepth) {
+            StackTraceElement[] stackTrace = thread.getStackTrace();
+            return maxDepth == -1 || maxDepth >= stackTrace.length ? stackTrace : Arrays.copyOfRange(stackTrace, 0, maxDepth);
+        }
+
+        private record Blocker(Object blockObject, Thread ownerThread) {
+        }
+
+        private static Blocker getBlockerInfo(Thread thread) {
+            Object blocker = LockSupport.getBlocker(thread);
+
+            if (blocker instanceof JavaMonitor javaMonitor) {
+                return new Blocker(
+                                javaMonitor.getBlockedObject(),
+                                SubstrateThreadMXBean.getThreadById(javaMonitor.getOwnerThreadId()));
+            } else if (blocker instanceof AbstractOwnableSynchronizer synchronizer) {
+                return new Blocker(synchronizer,
+                                SubstrateUtil.cast(synchronizer, Target_java_util_concurrent_locks_AbstractOwnableSynchronizer.class)
+                                                .getExclusiveOwnerThread());
+            }
+            return new Blocker(blocker, null);
+        }
+
+        private static Object[] getLockedSynchronizers(Thread thread) {
+            EconomicSet<AbstractOwnableSynchronizer> locks = JMXMonitoring.getThreadLocks(thread);
+            if (locks != null) {
+                synchronized (locks) {
```

Comment:
I believe there is still a race here. I don't think it's enough to only synchronize here when we're reading the locks. We also need to synchronize when we modify the set of locks . For example, if the set of locks is modified by the target thread in between when we assign `lockObjects` a size and when we finish iterating the set, then we may try to access an index that doesn't exist.

Reply:
Hi @roberttoyonaga, indeed there was such possibility.
So, only one thread is adding or removing locks - it's the thread that owns them. Another thread is reading these values for reporting here. I did a small change here to use an iterator over the locks. This way I'm sure that any modifications of the underlying collection won't affect me when I'm reading the collection (and I think I shouldn't block the writes while reading it).

Reply:
This is better than before, but doesn't prevent the race. While iterating the list here, the owning thread may concurrently remove items from the list. Will this cause problems?

Reply:
Hi @roberttoyonaga,
when the values get removed, they are set to null (in an array), and on iterating, the null values are skipped. However, there might be a problem when the size shrinks, if that happens in between `hasNext()` and `next()` calls.
I've made a change. It's really convenient to continue using the `EconomicSet` since I can provide it with equivalence strategy that uses `System.identityHashCode`. So I decided to go for external synchronization. I'm not using the `ReadWriteLock` here because that adds an extra lock that the thread obtains while it's already busy adding a lock, so it will loop trying to add more lock entries. So the monitor it is.

I've also made a change for the monitors collection, where I'm using now `ConcurrentLinkedDeque`.

Reply:
> So the monitor it is

I think your new approach has solved the concurrency problem here. However, I think it's possible to easily optimize this with more fine-grained locking. You could maybe synchronize on the lock collection itself (`Target_java_lang_Thread#locks`) so multiple threads can work simultaneously on different collections.  Is there a reason you are using a [global lock](https://github.com/oracle/graal/pull/7945/files#diff-75f469d0228df9c99ef350ab92b31c4beccc00802f61903041bf45682a4b65c7R101) instead?

> I've also made a change for the monitors collection, where I'm using now ConcurrentLinkedDeque.

What is the purpose of this change? Changing to a `ConcurrentLinkedDeque` does not make the construction of the list copy operation atomic ([`return List.copyOf(targetThread.monitors);`](https://github.com/oracle/graal/pull/7945/files#diff-75f469d0228df9c99ef350ab92b31c4beccc00802f61903041bf45682a4b65c7R610)). So I still have the [same comment as before.](https://github.com/oracle/graal/pull/7945#discussion_r1463860189)

Reply:
Indeed, I shouldn't be using the global lock. Fixed that.

Regarding the `ConcurrentLinkedDeque`, I should have used the Iterator that it provides instead of `List.copyOf`. According to the documentation, this Iterator is weakly consistent:

> - they may proceed concurrently with other operations
> - they will never throw `ConcurrentModificationException`
> - they are guaranteed to traverse elements as they existed upon construction exactly once, and may (but are not guaranteed to) reflect any modifications subsequent to construction.